app/models/manageiq/providers/ibm_cloud/vpc/cloud_manager/logging_mixin.rb

```diff
@@ -0,0 +1,85 @@
+# frozen_string_literal: true
+
+# Encapsulate logging methods into a single mixin.
+module ManageIQ::Providers::IbmCloud::VPC::CloudManager::LoggingMixin
+  # A logger wrapper that can be used as a general interface during dev or production.
+  # During development use the defined $ibm_cloud_log which prints to ibm_cloud.log.
+  # In production use the defined _log which prints to general evm log and should be captured in containers.
+  class LoggingWrapper
+    # Define a short-lived object for this method.
+    # @param method_name [String] The name of the method calling this logger.
+    # @return [void]
+    def initialize(method_name)
+      @method_name = method_name
+    end
+
+    # Send a info message to the configured logger.
+    # @param msg [String] A string message to print.
+    # @return [void]
+    def info(msg)
+      parent_logger.info(format_message(msg))
+    end
+
+    # Send a debug message to the configured logger.
+    # @param msg [String] A string message to print.
+    # @return [void]
+    def debug(msg)
+      parent_logger.debug(format_message(msg))
+    end
+
+    # Send a warn message to the configured logger.
+    # @param msg [String] A string message to print.
+    # @return [void]
+    def warn(msg)
+      parent_logger.warn(format_message(msg))
+    end
+
+    # Send a warn message to the configured logger.
+    # @param msg [String] A string message to print.
+    # @return [void]
+    def error(msg)
+      parent_logger.error(format_message(msg))
+    end
+
+    # Log the backtrace for an exception.
+    # @param exception [Exception] The exception object.
+    # @param context_msg [String] A message to be printed as an error that provides context for the exception.
+    # @param re_raise [Boolean] Whether to raise the original error again.
+    # @param send_return [] The object to send as a return value.
+    # @return [void]
+    def log_backtrace(exception, context_msg: '', re_raise: true, send_return: nil)
+      parent_logger.error(format_message(context_msg)) unless context_msg.length.zero?
+
+      parent_logger.log_backtrace(exception)
+      raise if re_raise
+
+      send_return
+    end
+
+    private
+
+    # Append the callers class and method_name to the message.
+    # @param msg [String] The message string.
+    # @return [String]
+    def format_message(msg)
+      "#{@method_name} #{msg}"
+    end
+
+    # Determine which logger to use for this call.
+    # @return [Vmdb::Loggers::ProviderSdkLogger, Vmdb::LogProxy]
+    def parent_logger
+      development? ? $ibm_cloud_log : _log
```

Comment:
This throws an exception if `development?` returns false

Reply:
Really? I thought I tested that. No, I meant _log. Like in `_log.info("IBM SERVER PROVISIONING OPTIONS: #{json_clone_options}")`. Could it be that the REPL doesn't have it set, but a running server does? I think, II found I could only test it as a server and not from the rails console.

Reply:
> Could it be that the REPL doesn't have it set, but a running server does? I think, if I remember correctly, I found I could only test it as a server and not from the rails console.

Yeah if you want to call `_log` you have to `include Vmdb::Logging`  Often times this is done in one of the base classes for you but since you're defining your own `LoggingWrapper` class it is up to you to include this.

Even if this worked with the full server running that indicates there is a dependency load ordering issue which will come back to bite us.

> Is there a log method? Where does that go?

```
=> #<Method: main.log() /home/grare/adam/.gem/gems/yard-0.9.26/lib/yard/globals.rb:20>
>> log
=> #<YARD::Logger:0x0000564a04a7ff90 @level=2, @progname=nil, @default_formatter=#<Logger::Formatter:0x0000564a04a7fef0 @datetime_format=nil>, @formatter=#<Method: YARD::Logger#format_log(sev, _time, _prog, msg) /home/grare/adam/.gem/gems/yard-0.9.26/lib/yard/logging.rb:201>, @logdev=#<IO:<STDOUT>>, @show_backtraces=true, @show_progress=false, @warned=false, @progress_indicator=0, @mutex=#<Thread::Mutex:0x0000564a04a7fd88>, @progress_msg=nil, @progress_last_update=2021-05-04 09:00:25.637672949 -0400>
```

So that's not what we want here


Running in production:
```
RAILS_ENV=production rails c
>> vpc_provision = ManageIQ::Providers::IbmCloud::VPC::CloudManager::Provision.new
>> vpc_provision.logger("foo").info("hello, world")
Traceback (most recent call last):
        2: from /home/grare/adam/src/manageiq/manageiq-providers-ibm_cloud/app/models/manageiq/providers/ibm_cloud/vpc/cloud_manager/logging_mixin.rb:20:in `info'
        1: from /home/grare/adam/src/manageiq/manageiq-providers-ibm_cloud/app/models/manageiq/providers/ibm_cloud/vpc/cloud_manager/logging_mixin.rb:71:in `parent_logger'
NameError (undefined local variable or method `_log' for #<ManageIQ::Providers::IbmCloud::VPC::CloudManager::LoggingMixin::LoggingWrapper:0x00005624ae5dc6d0>)
Did you mean?  log
```

And if we include `Vmdb::Logger`:
```
   class LoggingWrapper
+    include Vmdb::Logging
+
     # Define a short-lived object for this method.
```

```
RAILS_ENV=production rails c
>> vpc_provision = ManageIQ::Providers::IbmCloud::VPC::CloudManager::Provision.new
>> vpc_provision.logger("foo").info("hello, world")
=> true
```